async_timeout/__init__.py

```diff
@@ -206,17 +206,34 @@ def _do_enter(self) -> None:
         self._state = _State.ENTER
         self._reschedule()
 
-    def _do_exit(self, exc_type: Optional[Type[BaseException]]) -> None:
+    def _do_exit(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+    ) -> None:
         if exc_type is asyncio.CancelledError and self._state == _State.TIMEOUT:
-            self._timeout_handler = None
-            raise asyncio.TimeoutError
-        # timeout has not expired
-        self._state = _State.EXIT
+            skip = False
+            if sys.version_info >= (3, 9):
+                # Analyse msg
+                assert exc_val is not None
+                if not exc_val.args or exc_val.args[0] != id(self):
+                    skip = True
+            if not skip:
+                if sys.version_info >= (3, 11):
+                    asyncio.current_task().uncancel()
+                raise asyncio.TimeoutError
```

Comment:
Following what Guido was saying, I think this needs to check to check the cancelled count, maybe something like:
```suggestion
            if sys.version_info >= (3, 11):
                if asyncio.current_task().uncancel() == 0:
                    self._timeout_handler = None
                    raise asyncio.TimeoutError()
                else:
                    raise asyncio.TimeoutError()
```

I think this is all that's needed, as the `== _State.TIMEOUT` tells us that we initiated a cancel, so there's no need for the cancel message anymore. Would be good to test this out against those previous tests before wrapping up the cpython discussion.